File: app/src/main/java/wangdaye/com/geometricweather/common/ui/decotarions/ListDecoration.java
package wangdaye.com.geometricweather.common.ui.decotarions;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Px;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;

import wangdaye.com.geometricweather.common.utils.DisplayUtils;
import wangdaye.com.geometricweather.common.utils.managers.ThemeManager;

/**
 * List decoration.
 * */

public class ListDecoration extends RecyclerView.ItemDecoration {

    private final Paint mPaint;
    private @Px final int mDividerDistance;

    public ListDecoration(Context context) {
        this(context, ThemeManager.getInstance(context).getLineColor(context));
    }

    public ListDecoration(Context context, @ColorInt int color) {
        mDividerDistance = (int) DisplayUtils.dpToPx(context, 1);

        mPaint = new Paint();
        mPaint.setColor(color);
        mPaint.setStyle(Paint.Style.STROKE);
        mPaint.setStrokeWidth(mDividerDistance);
    }

    @Override
    public void onDraw(@NonNull Canvas c, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        for (int i = 0; i < parent.getChildCount(); i ++){
            View child = parent.getChildAt(i);
            c.drawLine(
                    child.getLeft(),
                    child.getBottom() + mDividerDistance / 2f,
                    child.getRight(),
                    child.getBottom() + mDividerDistance / 2f,
                    mPaint
            );
        }
    }

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view,
                               @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        outRect.set(0, 0, 0, mDividerDistance);
    }
}

File: app/src/main/java/org/breezyweather/common/ui/decorations/ListDecoration.kt
package org.breezyweather.common.ui.decorations

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.view.View
import androidx.annotation.ColorInt
import androidx.annotation.Px
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration
import org.breezyweather.common.extensions.dpToPx

/**
 * List decoration.
 */
class ListDecoration(context: Context, @ColorInt colorP: Int) : ItemDecoration() {
    private val mPaint: Paint

    @Px
    private val mDividerDistance: Int

    init {
        mDividerDistance = context.dpToPx(1f).toInt()
        mPaint = Paint().apply {
            color = colorP
            style = Paint.Style.STROKE
            strokeWidth = mDividerDistance.toFloat()
        }
    }

    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        for (i in 0 until parent.childCount) {
            val child = parent.getChildAt(i)
            c.drawLine(
                child.left.toFloat(),
                child.bottom + mDividerDistance / 2f,
                child.right.toFloat(),
                child.bottom + mDividerDistance / 2f,
                mPaint
            )
        }
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        outRect.set(0, 0, 0, mDividerDistance)
    }

    @get:ColorInt
    var color: Int
        get() = mPaint.color
        set(color) {
            mPaint.color = color
        }
}
